rs/messaging/src/routing/stream_handler/tests.rs

```diff
@@ -3468,281 +2870,599 @@ fn process_stream_slices_canister_migration_in_both_subnets_success() {
-        slice.header_mut().set_end(end.into());
+/// Returns an iterator over messages in a stream over the `stream_index_range`.
+///
+/// Panics if any of the messages in `stream_index_range` does not exist.
+fn messages_in_stream(
+    opt_stream: Option<&Stream>,
+    stream_index_range: std::ops::RangeInclusive<u64>,
+) -> impl Iterator<Item = &RequestOrResponse> {
+    match opt_stream {
+        // Not the `unwrap()`s here ensure that the code panics when the whole range was not
+        // available rather than just silently terminating the iterator prematurely. This behaviour
+        // is important to ensure we actually test as intended.
+        Some(stream) => stream_index_range
+            .map(|stream_index| stream.messages().get(stream_index.into()).unwrap()),
+        None => unreachable!(),
     }
-    slice.header_mut().set_flags(config.flags);
-    slice
 }
 
-fn get_request_at(queue: &StreamIndexedQueue<RequestOrResponse>, stream_index: u64) -> &Request {
-    match queue.get(stream_index.into()) {
-        Some(RequestOrResponse::Request(request)) => request,
-        _ => unreachable!(),
+/// Returns an iterator over messages in a stream slice over the `stream_index_range`.
+///
+/// Panics if any of the messages in `stream_index_range` does not exist.
+fn messages_in_slice(
+    opt_slice: Option<&StreamSlice>,
+    stream_index_range: std::ops::RangeInclusive<u64>,
+) -> impl Iterator<Item = &RequestOrResponse> {
+    match opt_slice {
+        // Not the `unwrap()`s here ensure that the code panics when the whole range was not
+        // available rather than just silently terminating the iterator prematurely. This behaviour
+        // is important to ensure we actually test as intended.
+        Some(slice) => stream_index_range
+            .map(|stream_index| slice.messages().unwrap().get(stream_index.into()).unwrap()),
+        None => unreachable!(),
     }
 }
 
-fn get_response_at(queue: &StreamIndexedQueue<RequestOrResponse>, stream_index: u64) -> &Response {
-    match queue.get(stream_index.into()) {
-        Some(RequestOrResponse::Response(response)) => response,
-        _ => unreachable!(),
+/// Pushes the messages yielded by `iter` into the `state`.
+fn push_inputs<'a>(
+    state: &mut ReplicatedState,
+    iter: impl Iterator<Item = &'a RequestOrResponse>,
+) -> Result<(), (StateError, RequestOrResponse)> {
+    for msg in iter {
+        state.push_input(msg.clone(), &mut (i64::MAX / 2))?;
+    }
+    Ok(())
+}
+
+/// Instructions for building various kinds of messages; essentially a wrapper for
+/// `RequestBuilder`, `ResponseBuilder`.
+#[allow(clippy::enum_variant_names)]
+#[derive(Copy, Clone)]
+enum MessageBuilder {
+    // `(sender, receiver)`.
+    BuildRequest(CanisterId, CanisterId),
+    // `(respondent, originator)`
+    BuildResponse(CanisterId, CanisterId),
+    // `(respondent, originator, reason)`
+    BuildRejectResponse(CanisterId, CanisterId, RejectReason),
+}
```

Comment:
Why not simply name the variants `Request`, `Response` (or `DataResponse`) and `RejectResponse`? I don't think we ever `use RequestRoResponse::*` or similar in this file, so there should be no conflict.

Reply:
I figured it might be confusing since I'm using `StreamConfig` with instructions on what kind of message we want in the setup stage and then also with actual messages in the 'construct the expected state' stage. Also there are a number of tests and functions that do use `Request` and `Response` but this can be resolved by using `ic_types::messages::Request` instead.

I guess it's not really possible to mistake `Request(*LOCAL_CANISTER, *LOCAL_CANISTER)` for an actual messages anyway so here you go.